lock-api/src/main/java/com/palantir/lock/client/CommitTimestampGetter.java

```diff
@@ -0,0 +1,105 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.lock.client;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.function.Consumer;
+import java.util.stream.Collectors;
+import java.util.stream.LongStream;
+
+import org.immutables.value.Value;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.Streams;
+import com.palantir.atlasdb.autobatch.Autobatchers;
+import com.palantir.atlasdb.autobatch.BatchElement;
+import com.palantir.atlasdb.autobatch.DisruptorAutobatcher;
+import com.palantir.atlasdb.futures.AtlasFutures;
+import com.palantir.atlasdb.timelock.api.GetCommitTimestampsResponse;
+import com.palantir.lock.v2.LockToken;
+import com.palantir.lock.watch.ImmutableTransactionUpdate;
+import com.palantir.lock.watch.LockWatchEventCache;
+import com.palantir.lock.watch.TransactionUpdate;
+
+final class CommitTimestampGetter implements AutoCloseable {
+    private final DisruptorAutobatcher<Request, Long> autobatcher;
+
+    private CommitTimestampGetter(DisruptorAutobatcher<Request, Long> autobatcher) {
+        this.autobatcher = autobatcher;
+    }
+
+    public static CommitTimestampGetter create(LockLeaseService leaseService, LockWatchEventCache cache) {
+        DisruptorAutobatcher<Request, Long> autobatcher = Autobatchers
+                .independent(consumer(leaseService, cache))
+                .safeLoggablePurpose("get-commit-timestamp")
+                .build();
+        return new CommitTimestampGetter(autobatcher);
+    }
+
+    public long getCommitTimestamp(long startTs, LockToken commitLocksToken) {
+        return AtlasFutures.getUnchecked(autobatcher.apply(ImmutableRequest.builder()
+                .startTs(startTs)
+                .commitLocksToken(commitLocksToken)
+                .build()));
+    }
+
+    @VisibleForTesting
+    static Consumer<List<BatchElement<Request, Long>>> consumer(LockLeaseService leaseService,
+            LockWatchEventCache cache) {
+        return batch -> {
+            int count = batch.size();
+            List<Long> commitTimestamps = new ArrayList<>();
+            while (commitTimestamps.size() < count) {
+                GetCommitTimestampsResponse response = leaseService.getCommitTimestamps(cache.lastKnownVersion(),
+                        count - commitTimestamps.size());
+                commitTimestamps.addAll(process(batch.subList(commitTimestamps.size(), count), response, cache));
+            }
+
+            for (int i = 0; i < count; i++) {
+                batch.get(i).result().set(commitTimestamps.get(i));
+            }
```

Comment:
non-actionable: This is a little tricky, but makes sense (the cache updates itself so in the next batch you only ask for the updates from that point onwards).

Reply:
Yeah, this stitching is annoying, but I believe it's correct.